File: cpp/tests/test_rrt_connect.cpp
#include <stdexcept>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/sampling/rrt_connect.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open 10x10 grid (5m x 5m world).
std::vector<std::string> open_rows(int n = 10) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

std::string rrt_connect_cfg() {
  return test::write_temp(
      "rrt_connect.yaml",
      "algorithm: rrt_connect\ncategory: global_planning\nparams:\n"
      "  - {name: max_iterations, type: int, default: 4000, min: 1, max: 200000, description: n}\n"
      "  - {name: step_size, type: float, default: 0.5, min: 0.01, max: 100, description: eta}\n"
      "  - {name: goal_tolerance, type: float, default: 0.3, min: 0.0, max: 100, description: t}\n"
      "  - {name: seed, type: int, default: 7, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid path reaching the goal --------------------------------------------

TEST(RrtConnect, FindsValidPath) {
  auto g = test::make_grid(open_rows());
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::RrtConnectPlanner p(core::ParamSet::from_yaml(rrt_connect_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  ASSERT_GE(r.path.size(), 2u);
  // Both trees are rooted at exact states, so the spliced path pins to start/goal.
  EXPECT_NEAR(r.path.front().x, start.x, 1e-9);
  EXPECT_NEAR(r.path.front().y, start.y, 1e-9);
  EXPECT_NEAR(r.path.back().x, goal.x, 1e-9);
  EXPECT_NEAR(r.path.back().y, goal.y, 1e-9);
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1]))
        << "path segment must be collision-free";
  }
  EXPECT_GT(r.cost, 0.0);
}

// (b) no-path case: a full wall separates start from goal ---------------------

TEST(RrtConnect, NoPathAcrossFullWall) {
  auto rows = open_rows();
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::RrtConnectPlanner p(core::ParamSet::from_yaml(rrt_connect_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  EXPECT_EQ(r.cost, 0.0);
}

// (c) param validation failure ------------------------------------------------

TEST(RrtConnect, BadStepSizeThrows) {
  std::string bad = test::write_temp(
      "rrt_connect.yaml",
      "algorithm: rrt_connect\ncategory: global_planning\nparams:\n"
      "  - {name: step_size, type: float, default: 200.0, min: 0.01, max: 100.0, description: over max}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

File: python/tests/test_rrt_connect.py
"""RRT-Connect: bidirectional trees meet on open space, fail when walled off."""

from __future__ import annotations

from pathlib import Path

import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning import RRTConnect
from navigation.maps.occupancy_grid import OccupancyGrid2D


def _capped_config(tmp_path: Path, max_iter: int) -> ParamSet:
    """Real rrt_connect config with a smaller budget so the no-path test stays fast."""
    doc = yaml.safe_load((CONFIG_DIR / "rrt_connect.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "max_iterations":
            entry["default"] = max_iter
    out = tmp_path / "rrt_connect.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_on_open_grid(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.5, 0.5), (9.0, 9.0)
    params = _capped_config(tmp_path, max_iter=2000)
    result = RRTConnect(params).plan(grid, start, goal)
    assert result.success
    # Both roots are exact tree nodes, so the spliced path pins to start and goal.
    assert result.path[0] == pytest.approx(start)
    assert result.path[-1] == pytest.approx(goal)
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: the two trees can never connect.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _capped_config(tmp_path, max_iter=400)
    result = RRTConnect(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_out_of_range_step_size_rejected(tmp_path: Path) -> None:
    # step_size above its declared max must fail at load time, not silently clamp.
    bad = write_config(
        tmp_path / "rrt_connect.yaml",
        "rrt_connect",
        [{"name": "step_size", "type": "float", "default": 200.0, "min": 0.01, "max": 100.0,
          "description": "over max"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)
